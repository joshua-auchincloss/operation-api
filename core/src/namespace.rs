use std::collections::BTreeMap;

use convert_case::Casing;

use crate::{Definitions, Field, FieldOrRef, Ident, Operation, Struct, generate::LanguageTrait};

pub trait OfNamespace {
    const NAMESPACE: &'static str;
}

#[macro_export]
macro_rules! namespace {
    ($ns: literal {
        $($t: path), + $(,)?
    }) => {
        $(
            impl $crate::namespace::OfNamespace for $t {
                const NAMESPACE: &'static str = $ns;
            }
        )*
    };
}

#[derive(serde::Deserialize, serde::Serialize, Debug)]
pub struct Namespace {
    pub name: Ident,

    pub fields: BTreeMap<Ident, Field<Ident>>,
    pub ops: BTreeMap<Ident, Operation>,
    pub defs: BTreeMap<Ident, Struct>,
}

impl Namespace {
    pub fn new<I: Into<Ident>>(name: I) -> Self {
        Self {
            name: name.into(),
            fields: Default::default(),
            ops: Default::default(),
            defs: Default::default(),
        }
    }

    pub fn with_definitions(
        &mut self,
        defs: Vec<Definitions>,
    ) -> crate::Result<()> {
        for def in defs {
            self.with_definition(def)?;
        }
        Ok(())
    }

    pub fn with_definition(
        &mut self,
        def: Definitions,
    ) -> crate::Result<()> {
        match def {
            Definitions::FieldV1(field) => {
                unique_ns_def(
                    &mut self.fields,
                    field.name.clone(),
                    &self.name,
                    field,
                    "field",
                )?;
            },
            Definitions::StructV1(def) => {
                unique_ns_def(&mut self.defs, def.name.clone(), &self.name, def, "struct")?;
            },
            Definitions::OperationV1(op) => {
                unique_ns_def(&mut self.ops, op.name.clone(), &self.name, op, "operation")?;
            },
        }

        Ok(())
    }

    pub fn resolve_field(
        &self,
        name: &Ident,
    ) -> crate::Result<&Field<Ident>> {
        self.fields.get(name).ok_or_else(|| {
            crate::Error::NameNotFound {
                name: name.clone(),
                ns: self.name.clone(),
            }
        })
    }

    pub fn resolve_field_types(&mut self) -> crate::Result<()> {
        for (def_name, def) in self.defs.clone().iter() {
            let mut swap = vec![];
            for (field_name, field) in &*def.fields {
                match field {
                    FieldOrRef::Field(..) => {},
                    FieldOrRef::Ref { to } => {
                        if let Some(local_ref) = def.fields.get(&to) {
                            swap.push((field_name.clone(), local_ref.clone()));
                        } else {
                            swap.push((
                                field_name.clone(),
                                FieldOrRef::Field(self.resolve_field(to)?.clone().into()),
                            ));
                        }
                    },
                }
            }

            for (name, field) in swap {
                self.defs
                    .get_mut(&def_name)
                    .unwrap()
                    .fields
                    .insert(name, field);
            }
        }

        Ok(())
    }

    pub fn check(&mut self) -> crate::Result<()> {
        self.resolve_field_types()?;
        Ok(())
    }

    pub fn normalized_path<L: LanguageTrait>(&self) -> String {
        self.name
            .to_string()
            .replace(".", "_")
            .to_case(L::file_case())
    }
}

#[inline]
fn unique_ns_def<T>(
    sources: &mut BTreeMap<Ident, T>,
    name: Ident,
    ns: &Ident,
    def: T,
    tag: &'static str,
) -> crate::Result<()> {
    match sources.insert(name.clone(), def) {
        Some(..) => {
            Err(crate::Error::NamespaceConflict {
                ns: ns.clone(),
                name,
                tag,
            })
        },
        None => Ok(()),
    }
}
