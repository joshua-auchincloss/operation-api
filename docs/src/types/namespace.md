# namespace

Declares a namespace, either for the file or the spanned contents.

Syntax:

```pld
# some foo/baz.pld
namespace foo;
```

```pls
namespace foo {
    type Abc = str | i32;
};
```
