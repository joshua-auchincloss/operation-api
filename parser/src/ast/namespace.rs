use crate::{
    SpannedToken, Token,
    ast::AstStream,
    bail_unchecked,
    defs::Spanned,
    tokens::{self, Brace, LBraceToken, Parse, Peek, ToTokens, brace},
};

pub struct Namespace {
    pub kw: SpannedToken![namespace],
    pub name: SpannedToken![ident],
}

impl Parse for Namespace {
    fn parse(stream: &mut crate::tokens::TokenStream) -> Result<Self, crate::tokens::LexingError> {
        Ok(Self {
            kw: stream.parse()?,
            name: stream.parse()?,
        })
    }
}

impl Peek for Namespace {
    fn peek(stream: &tokens::TokenStream) -> bool {
        stream.peek::<Token![namespace]>() && !stream.peek::<SpannedNamespace>()
    }
}

impl ToTokens for Namespace {
    fn write(
        &self,
        tt: &mut crate::fmt::Printer,
    ) {
        tt.write(&self.kw);
        tt.space();
        tt.write(&self.name);
    }
}

pub struct SpannedNamespace {
    pub kw: SpannedToken![namespace],
    pub name: SpannedToken![ident],
    pub brace: Brace,
    pub ast: Spanned<AstStream>,
}

impl Peek for SpannedNamespace {
    fn peek(stream: &tokens::TokenStream) -> bool {
        let mut fork = stream.fork();

        let _: SpannedToken![namespace] = bail_unchecked!(fork.parse(); false);
        let _: SpannedToken![ident] = bail_unchecked!(fork.parse(); false);

        fork.peek::<LBraceToken>()
    }
}

impl Parse for SpannedNamespace {
    fn parse(stream: &mut crate::tokens::TokenStream) -> Result<Self, crate::tokens::LexingError> {
        let mut braced;
        Ok(Self {
            kw: stream.parse()?,
            name: stream.parse()?,
            brace: brace!(braced in stream),
            ast: braced.parse()?,
        })
    }
}

impl ToTokens for SpannedNamespace {
    fn write(
        &self,
        tt: &mut crate::fmt::Printer,
    ) {
        tt.write(&self.kw);
        tt.space();
        tt.write(&self.name);
        tt.space();
        tt.open_block();
        tt.write(&self.ast);
        tt.close_block();
    }
}
