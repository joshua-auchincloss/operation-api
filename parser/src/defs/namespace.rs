use crate::defs::*;

use pest::iterators::Pairs;

use crate::parser::Rule;

#[derive(Debug, Clone, bon::Builder)]
pub struct NamespaceDef<V> {
    pub comment: String,
    pub ident: Vec<Ident>,
    pub meta: Vec<Meta>,
    pub version: V,
}

impl<V: Default> FromInner for NamespaceDef<V> {
    fn from_inner(pairs: Pairs<crate::parser::Rule>) -> crate::Result<Self> {
        let mut comment = String::new();
        for p in pairs {
            match p.as_rule() {
                Rule::singleline_comment | Rule::multiline_comment => {
                    comment += &take_comment(Pairs::single(p));
                },
                Rule::ident => {
                    let sp_ident = Ident::from_pair_span(p)?;
                    return Ok(Self::builder()
                        .ident(sp_ident.value.qualified_path())
                        .comment(comment)
                        .meta(Vec::new())
                        .version(V::default())
                        .build());
                },
                _ => {},
            }
        }
        Err(crate::Error::def::<Self>(Rule::ident))
    }
}

impl<V> Commentable for NamespaceDef<V> {
    fn comment(
        &mut self,
        comment: String,
    ) {
        self.comment += &comment;
    }
}

impl<V: Default> FromPairSpan for NamespaceDef<V> {
    fn from_pair_span(pair: pest::iterators::Pair<'_, Rule>) -> crate::Result<Spanned<Self>> {
        let span = pair.as_span();
        let start = span.start();
        let end = span.end();
        let value = NamespaceDef::from_inner(pair.into_inner())
            .map_err(crate::Error::then_with_span(start, end))?;
        Ok(Spanned::new(start, end, value))
    }
}

pub type NamespaceSealed = NamespaceDef<usize>;
pub type NamespaceUnsealed = NamespaceDef<Option<usize>>;

impl NamespaceUnsealed {
    pub fn seal(
        self,
        file_version: usize,
    ) -> NamespaceSealed {
        NamespaceDef {
            comment: self.comment,
            ident: self.ident,
            meta: self.meta,
            version: self.version.unwrap_or(file_version),
        }
    }
}
