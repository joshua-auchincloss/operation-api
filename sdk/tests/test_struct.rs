#![allow(unused)]

use operation_api_core::namespace;
use operation_api_derives::Struct;
use operation_api_sdk::Defined;

#[derive(Struct)]
#[fields(version = 1)]
#[fields(describe(text = "Some struct"))]
pub struct BasicStruct {
    #[fields(describe(text = "field a"))]
    a: i32,
    b: Option<i32>,
    c: Vec<f32>,
    d: [[f32; 4]; 4],
}

#[derive(Struct)]
#[fields(version = 1)]
#[fields(describe(file = "../../samples/some-readme.md"))]
pub struct BasicStructWithReadme {
    #[fields(describe(text = "field a"))]
    a: i32,
}

namespace! {
    "abc.corp.test" {
        BasicStruct, BasicStructWithReadme
    }
}

fn smoke_basic<D: Defined, F: Fn(&String)>(
    out: &'static str,
    snap: F,
) {
    let ser = toml::to_string(D::definition()).unwrap();

    snap(&ser);

    std::fs::write(out, ser).unwrap();
}

#[test]
fn smoke_basic_with_readme() {
    smoke_basic::<BasicStructWithReadme, _>("../samples/test-struct-readme.toml", |ser| {
        insta::assert_snapshot!(ser)
    })
}

#[test]
fn smoke_basic_with_text() {
    smoke_basic::<BasicStruct, _>("../samples/test-struct-text.toml", |ser| {
        insta::assert_snapshot!(ser)
    })
}
