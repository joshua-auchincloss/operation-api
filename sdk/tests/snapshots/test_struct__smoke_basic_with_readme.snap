---
source: sdk/tests/test_struct.rs
expression: ser
---
"type = \"struct@v1\"\nname = \"BasicStructWithReadme\"\nnamespace = \"abc.corp.test\"\ndescription = \"\"\"\n# test\nthis is a struct description\n\"\"\"\nversion = 1\n\n[fields.a]\nname = \"a\"\nnamespace = \"abc.corp.test\"\ntype = \"i32\"\ndescription = \"field a\"\noptional = false\n"
